---
title: "FEV1/FVC Scores"
author: "Iain R. Konigsberg"
date: '`r Sys.Date()`'
output: 
  rmarkdown::html_document:
    code_folding: show
    toc: true
    toc_float: 
      collapsed: false
    df_print: paged
    number_sections: true
---

# Background

This Rmd contains code to evaluate FEV1/FVC scores. This Rmd generates Figure 2. 

***

# Set Up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = F)
knitr::opts_knit$set(root.dir = 'C:/Users/konigsbi/Desktop/ors_topmed/January/Results/predictions_copdgene/')

```

```{r message=F}
library(tidyverse)
library(dplyr)
library(ggpubr)
library(stringr)
library(tidyr)
library(sjPlot)
library(MASS)
library(rsq)
library(pROC)
library(glmnet)
library(glmnetUtils)
library(ggbeeswarm)
library(rstatix)

start = Sys.time()

theme_set(theme_bw())

set.seed(11)

setwd("C:/Users/konigsbi/Desktop/ors_topmed/January/Results/predictions_copdgene/")
files = dir()

model_select = function(mod){
  index = 1
  df = NULL
  for (x in 1:11){
    alpha = mod$alpha[x]
    moddy = mod$modlist[[x]]
    for (y in 1:length(moddy$lambda)){
      lambda = moddy$lambda[y]
      nonzero = moddy$nzero[y]
      measure = moddy$cvm[y]
      se = moddy$cvsd[y]
      df = rbind(df, c(alpha, index, lambda, nonzero, measure, se))
    }
    index = index + 1
  }
  df = as.data.frame(df)
  names(df) = c("alpha", "index", "lambda", "nonzero", "measure", "se")
  
  df = df[df$nonzero != 0, ]
  
  min_measure = df[df$measure == min(df$measure), ]$measure
  se_measure = df[df$measure == min(df$measure), ]$se
  sub_df = df[df$measure <= min_measure + se_measure, , drop = F]
  sub_df = sub_df[sub_df$alpha == max(sub_df$alpha), , drop = F]
  alpha_elastic = sub_df[sub_df$lambda == max(sub_df$lambda), ]$alpha
  lambda_elastic = sub_df[sub_df$lambda == max(sub_df$lambda), ]$lambda
  nonzero_elastic = sub_df[sub_df$lambda == max(sub_df$lambda), ]$nonzero
  alpha_index = sub_df[sub_df$lambda == max(sub_df$lambda), ]$index
  return(c(alpha_index, lambda_elastic, alpha_elastic))
}

```


```{r}
pd = read.csv("C:/Users/konigsbi/Desktop/ors_topmed/Data/COPDGene/pd_main_overlap.csv", row.names = 1)
pd$copd_P2 = NA
pd[!is.na(pd$FEV1_FVC_post_P2) & pd$FEV1_FVC_post_P2 < .7, ]$copd_P2 = 1
pd[!is.na(pd$FEV1_FVC_post_P2) & pd$FEV1_FVC_post_P2 > .7, ]$copd_P2 = 0

rownames(pd) = pd$sid

#panel a: cor. 
met = readRDS("C:/Users/konigsbi/Desktop/ors_topmed/January/Results/predictions_copdgene/predictions_metabolite_elastic_FEV1_FVC_post_P2.rds")
prot = readRDS("C:/Users/konigsbi/Desktop/ors_topmed/January/Results/predictions_copdgene/predictions_protein_elastic_FEV1_FVC_post_P2.rds")
rna = readRDS("C:/Users/konigsbi/Desktop/ors_topmed/January/Results/predictions_copdgene/predictions_transcript_elastic_FEV1_FVC_post_P2.rds")

met$type = "Metabolite"
prot$type = "Protein"
rna$type = "Transcript"

met = merge(met, pd %>% dplyr::select(sid, FEV1_FVC_post_P2, copd_P2, Age_P2, ccenter_P2, race, gender, smoking_status_P2, ATS_PackYears_P2), by = "row.names")
prot = merge(prot, pd %>% dplyr::select(sid, FEV1_FVC_post_P2, copd_P2, Age_P2, ccenter_P2, race, gender, smoking_status_P2, ATS_PackYears_P2), by = "row.names")
rna = merge(rna, pd %>% dplyr::select(sid, FEV1_FVC_post_P2, copd_P2, Age_P2, ccenter_P2, race, gender, smoking_status_P2, ATS_PackYears_P2), by = "row.names")

dat = as.data.frame(rbind(met, prot, rna))

dat[!is.na(dat$copd_P2) & dat$copd_P2 == 0, ]$copd_P2 = "Control"
dat[!is.na(dat$copd_P2) & dat$copd_P2 == 1, ]$copd_P2 = "Case"
dat = dat[!is.na(dat$copd_P2), ]


#make sure i have covars. 
p_met = signif(summary(lm(copd_P2 ~ s1 + gender + race + Age_P2 + ccenter_P2 + smoking_status_P2 + ATS_PackYears_P2, met))[[4]]["s1", "Pr(>|t|)"], 3)
p_prot = signif(summary(lm(copd_P2 ~ s1 + gender + race + Age_P2 + ccenter_P2 + smoking_status_P2 + ATS_PackYears_P2, prot))[[4]]["s1", "Pr(>|t|)"], 3)
p_rna = signif(summary(lm(copd_P2 ~ s1 + gender + race + Age_P2 + ccenter_P2 + smoking_status_P2 + ATS_PackYears_P2, rna))[[4]]["s1", "Pr(>|t|)"], 3)

stat.test <- tibble::tribble(
  ~group1, ~group2, ~p.adj,  ~y.position, ~type,
  "Case",    "Control",    p_met,  0.97,          "Metabolite",
  "Case",    "Control",    p_prot, 0.97,          "Protein",
  "Case",    "Control",    p_rna,   0.97,        "Transcript"
)


a = ggplot(dat, aes(copd_P2, s1)) +
  geom_violin(aes(fill = copd_P2), alpha = 0.4) + 
  geom_boxplot(aes(fill = copd_P2), width = 0.45, outliers = F) +
  #geom_beeswarm(size = 0.4) +
  facet_wrap(~type) +
  theme(legend.position = "none", strip.background = element_blank()) +
  labs(x = "", y = "FEV1/FVC Score") +
  scale_fill_manual(values = c("#F8766D", "#619CFF")) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  ggprism::add_pvalue(stat.test, tip.length = 0.02)



#panel b: features
met = readRDS("C:/Users/konigsbi/Desktop/ors_topmed/January/Results/models_copdgene/model_metabolite_FEV1_FVC_post_P2.rds")
prot = readRDS("C:/Users/konigsbi/Desktop/ors_topmed/January/Results/models_copdgene/model_protein_FEV1_FVC_post_P2.rds")
rna = readRDS("C:/Users/konigsbi/Desktop/ors_topmed/January/Results/models_copdgene/model_transcript_FEV1_FVC_post_P2.rds")

metnames = read.csv("C:/Users/konigsbi/Desktop/ors_topmed/Data/COPDGene/Metabolome/COPDGene_MetaboliteInformation_202303.csv")
protnames = read.csv("C:/Users/konigsbi/Desktop/ors_topmed/Data/COPDGene/Proteome/copdg_soma_5k_p2_metadata_Jun21.csv")
ctnames = read.csv("C:/Users/konigsbi/Desktop/ors_topmed/Data/COPDGene/Transcriptome/ctnames.csv")

#met
test = model_select(met)
coefs = as.data.frame(as.matrix(coef(met$modlist[[test[1]]], s = test[2])))
coefs = coefs[coefs$s1 != 0, , drop = F]
coefs = coefs[rownames(coefs) != "(Intercept)", , drop = F]
nrow(coefs) #249. 
coefs = coefs[order(abs(coefs$s1), decreasing = T), , drop = F]
coefs = coefs[1:20, , drop = F]
coefs$name = NA
for (x in 1:nrow(coefs)){
  coefs$name[x] = metnames[metnames$metabid == rownames(coefs)[x], ]$CHEMICAL_NAME
}
coefs$type = "Metabolite"
met = coefs

#prot
test = model_select(prot)
coefs = as.data.frame(as.matrix(coef(prot$modlist[[test[1]]], s = test[2])))
coefs = coefs[coefs$s1 != 0, , drop = F]
coefs = coefs[rownames(coefs) != "(Intercept)", , drop = F]
nrow(coefs) #172
coefs = coefs[order(abs(coefs$s1), decreasing = T), , drop = F]
coefs = coefs[1:20, , drop = F]
coefs$name = NA
for (x in 1:nrow(coefs)){
  coefs$name[x] = protnames[protnames$seqID == rownames(coefs)[x], ]$Target
}
coefs$type = "Protein"
prot = coefs

#trans
test = model_select(rna)
coefs = as.data.frame(as.matrix(coef(rna$modlist[[test[1]]], s = test[2])))
coefs = coefs[coefs$s1 != 0, , drop = F]
coefs = coefs[rownames(coefs) != "(Intercept)", , drop = F]
nrow(coefs) #249. 
coefs = coefs[order(abs(coefs$s1), decreasing = T), , drop = F]
coefs = coefs[1:20, , drop = F]
coefs$name = NA
for (x in 1:nrow(coefs)){
  coefs$name[x] = ctnames[ctnames$gene_id == rownames(coefs)[x], ]$gene_name
}
coefs$type = "Transcript"
rna = coefs


coefs = as.data.frame(rbind(met, prot, rna))
coefs$dir = NA
coefs[coefs$s1 > 0 , ]$dir = "Positive"
coefs[coefs$s1 < 0 , ]$dir = "Negative"


#shorten metab names manually
coefs[coefs$name == "sphingomyelin (d18:2/14:0, d18:1/14:1)*", ]$name = "sphingomyelin"
coefs[coefs$name == "1-(1-enyl-stearoyl)-2-linoleoyl-GPE (P-18:0/18:2)*", ]$name = "1-(1-enyl-stearoyl)-2-linoleoyl-GPE"
#coefs[coefs$name == "glycosyl-N-palmitoyl-sphingosine (d18:1/16:0)", ]$name = "glycosyl-N-palmitoyl-sphingosine"
#coefs[coefs$name == "", ]$name = ""

coefs$dir = factor(coefs$dir, levels = c("Positive", "Negative"))

b = ggplot(coefs, aes(abs(s1), reorder(name, abs(s1)))) + 
  geom_bar(stat = "identity", aes(fill = dir)) +
  facet_wrap(~type, scales = "free") +
  labs(x = "Feature Weight", y = "", fill = "Direction") +
  theme(strip.background = element_blank()) +
  scale_fill_manual(values = c("#F8766D", "#619CFF")) +
  scale_x_continuous(n.breaks = 4)


#panel c: auc. 
met = readRDS("C:/Users/konigsbi/Desktop/ors_topmed/January/Results/predictions_copdgene/predictions_metabolite_elastic_FEV1_FVC_post_P2.rds")
prot = readRDS("C:/Users/konigsbi/Desktop/ors_topmed/January/Results/predictions_copdgene/predictions_protein_elastic_FEV1_FVC_post_P2.rds")
rna = readRDS("C:/Users/konigsbi/Desktop/ors_topmed/January/Results/predictions_copdgene/predictions_transcript_elastic_FEV1_FVC_post_P2.rds")


rownames(pd) = pd$sid
met = merge(met, pd, by = "row.names")
prot = merge(prot, pd, by = "row.names")
rna = merge(rna, pd, by = "row.names")
mod_null = glm(copd_P2 ~ Age_P2 + gender + race + smoking_status_P2 + ATS_PackYears_P2 + ccenter_P2, met, family = binomial) 
mod_met = glm(copd_P2 ~ s1 + Age_P2 + gender + race + smoking_status_P2 + ATS_PackYears_P2 + ccenter_P2, met, family = binomial) 
mod_prot = glm(copd_P2 ~ s1 + Age_P2 + gender + race + smoking_status_P2 + ATS_PackYears_P2 + ccenter_P2, prot, family = binomial) 
mod_rna = glm(copd_P2 ~ s1 + Age_P2 + gender + race + smoking_status_P2 + ATS_PackYears_P2 + ccenter_P2, rna, family = binomial) 

#gg
obj1 = roc(mod_null$y, mod_null$fitted.values, ci = T, col = "black")
obj2 = roc(mod_rna$y, mod_rna$fitted.values, add = T, ci = T, col = "#619CFF")
obj3 = roc(mod_met$y, mod_met$fitted.values, add = T, ci = T, col = "#F8766D")
obj4 = roc(mod_prot$y, mod_prot$fitted.values, add = T, ci = T, col = "#00BA38")

c = ggroc(list(Clinical = obj1, `TransRS + Clinical` = obj2, `MetRS + Clinical` = obj3, `ProtRS + Clinical` = obj4)) + 
  scale_color_manual(values = c("black", "#619CFF", "#F8766D", "#00BA38")) +
  labs(color = "", y = "Sensitivity", x = "Specificity") +
  geom_text(aes(0.02, 0.42, label = "AUC = 0.783                    "), hjust = 1, color = "black", size = 4) +
  geom_text(aes(0.02, 0.37, label = "AUC = 0.815; p = 4.72e-4"), hjust = 1, color = "#619CFF", size = 4) +
  geom_text(aes(0.02, 0.32, label = "AUC = 0.831; p = 1.02e-5"), hjust = 1, color = "#F8766D", size = 4) +
  geom_text(aes(0.02, 0.27, label = "AUC = 0.828; p = 4.54e-5"), hjust = 1, color = "#00BA38", size = 4) 
  
#DeLong pvals
roc.test(obj1, obj2) #rna
roc.test(obj1, obj3) #met
roc.test(obj1, obj4) #prot


png("C:/Users/konigsbi/Desktop/ors_topmed/April/Figures/Figure2.png", units = "in", res = 500, width = 10, height = 10)
one = cowplot::plot_grid(a, c, nrow = 1, labels = c("a.", "b."), rel_widths = c(1, 1))
two = cowplot::plot_grid(b, labels = c("c."))
cowplot::plot_grid(one, two, nrow = 2, rel_heights = c(.65, 1))
dev.off()

```

***

# Wrap-Up

```{r}
Sys.time() - start

sessionInfo()
```

