---
title: "SPIROMICS Evaluation"
author: "Iain R. Konigsberg"
date: '`r Sys.Date()`'
output: 
  rmarkdown::html_document:
    code_folding: show
    toc: true
    toc_float: 
      collapsed: false
    df_print: paged
    number_sections: true
---

# Background

This Rmd contains code to evaluate ORS in SPIROMICS participants. 

***

# Set Up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T)
knitr::opts_knit$set(root.dir = "C:/Users/konigsbi/Desktop/ors_topmed/March/Results/predictions_spiromics/")
```

IMPORTANT: 10_eval_mesa.Rmd has to be run first. 

```{r message=F}
library(dplyr)
library(ggpubr)
library(stringr)
library(tidyr)
library(sjPlot)
library(MASS)
library(rsq)
library(pROC)
library(data.table)

theme_set(theme_bw())

set.seed(11)

setwd("C:/Users/konigsbi/Desktop/ors_topmed/March/Results/predictions_spiromics/")
files = dir()

start = Sys.time()
```

need an rna pd, and a met/prot pd
```{r}
#old code from trans rep
rna = as.data.frame(fread("C:/Users/konigsbi/Desktop/ors_topmed/Data/SPIROMICS/Transcriptome/meyers_topmed_to3_rnaseq_1.rsem_genes_tpm.txt", header = T)) #takes seconds
pd_rna = as.data.frame(fread("C:/Users/konigsbi/Desktop/ors_topmed/Data/SPIROMICS/Transcriptome/SPIROMICS_phs001927_RNA-seq_SampleAttributesDS_20220122.txt")) #3067x21
pd_rna = pd_rna %>% dplyr::select(SAMPLE_ID, NWGC_ID, Collection_visit, Age_at_collection, Investigator_ID)
pd_rna = pd_rna[pd_rna$Collection_visit == "BASELINE", ]
att = as.data.frame(fread("C:/Users/konigsbi/Desktop/ors_topmed/Data/SPIROMICS/Transcriptome/SPIROMICS_phs001927_RNA-seq_SampleMappingDS_20220122.txt"))
pd_rna = merge(pd_rna, att, by = "SAMPLE_ID") #1840 x 6
pd = as.data.frame(fread("C:/Users/konigsbi/Desktop/ors_topmed/Data/SPIROMICS/90987/dbgap_submission/UnprocessedPheno20230210/SPIROMICS_I_TOPMed_CORE6_3_LONGITUDINAL_20230209.txt"))
pd = pd[pd$VISIT == "VISIT_1", ]
pd_rna = merge(pd_rna, pd, by.x = "SUBJECT_ID", by.y = "DBGAPID", all.x = T) 
rownames(rna) = rna$gene_id
rna = rna[,3:ncol(rna)]
rna = rna[, colnames(rna) %in% pd_rna$NWGC_ID, ]
dim(rna) 
rownames(rna) = make.unique(stringr::str_split(rownames(rna), "\\.", simplify = T)[,1])
rna = t(rna)
rna = log(rna + 1)

```
  
  
met and prot  
```{r}  
#old prot
prot = as.data.frame(fread("C:/Users/konigsbi/Desktop/ors_topmed/Data/SPIROMICS/Proteome/spiromics_website/AS136_FILTER_ADAT_SP1_LAD_231023.csv", header = T)) #takes seconds
dim(prot) #4475 x 7631
prot = prot[prot$VISIT == "VISIT1", ]
dim(prot) #2136
#samps in rows. features in cols. 
rownames(prot) = prot$LAD_ID
prot = prot[, 35:ncol(prot)]
prot$QTL_REMOVE_FLAG = NULL

pd = as.data.frame(fread("C:/Users/konigsbi/Desktop/ors_topmed/Data/SPIROMICS/CORE_6.6_CLINICAL_LAD4_20230206/CORE 6.6_CLINICAL_LAD4_20230206.csv")) #not split by visit....
dim(pd) #2316 x 708. 
prot = prot[, rownames(prot) %in% pd$LAD4, ]
dim(prot) #2136
prot = log(prot + 1)
colnames(prot) = gsub("seq_", "", colnames(prot))



#
#
#old met
met = as.data.frame(fread("C:/Users/konigsbi/Desktop/ors_topmed/Data/SPIROMICS/Metabolome/spiromics_website/AS096_Metab_BatNorm_SP1_LAD2207.csv", header = T))
rownames(met) = met$LAD4
met = met[, 3:ncol(met)]
met = met %>% mutate(across(c("_35":"_999926119"), ~replace_na(., median(., na.rm=TRUE))))
map = as.data.frame(fread("C:/Users/konigsbi/Desktop/ors_topmed/Data/COPDGene/Metabolome/COPDGene_MetaboliteInformation_202303.csv", header = T))
map$CHEM_ID = paste0("_", map$CHEM_ID)
for (x in 1:ncol(met)){
  if (colnames(met)[x] %in% map$CHEM_ID){
    colnames(met)[x] = map[map$CHEM_ID == colnames(met)[x], ]$metabid
  }
}
met = met[rownames(met) %in% pd$LAD4, ]
dim(met) #2177
met = log(met + 1)

#add ct to pd
ct = as.data.frame(data.table::fread("C:/Users/konigsbi/Desktop/ors_topmed/Data/SPIROMICS/CORE_CT_6_3_LAD_20221028/Core6_3_ct_master_sp1_lad_221028.csv"))
ct = ct[ct$VISIT == "VISIT_1", ]
ct = ct[ct$ANALYSIS_STATUS == "Passed", ] 
ct = ct[ct$LEFT_ACCEPTED_UNRELIABLE == "", ]
ct$selecter = paste0(ct$LAD4, ct$SEX)
ct = ct[ct$selecter != "244504M", ]
pd = merge(pd, ct, by = "LAD4", all.x = T)

dim(prot)
dim(met)
dim(pd)

dim(rna)
dim(pd_rna)


```


add preds to pd
```{r}
dir = "C:/Users/konigsbi/Desktop/ors_topmed/March/Results/predictions_spiromics/"
files = dir(dir)
files = paste0(dir, files)

#files = files[-grep("GasTrap", files)]

rna_files = files[grep("transcript", files)]
files = files[-grep("transcript", files)]

x = files[1]
comb = data.frame()
for (x in files){
  #x = files[1]
  cur = readRDS(x)
  names(cur) = gsub("model_", "", colnames(cur))
  cur = cur[!is.na(cur$Row.names), ]
  rownames(cur) = cur$Row.names
  comb = merge(comb, cur, by = "row.names", all = T)
  rownames(comb) = comb$Row.names
  comb$Row.names = NULL
}     

comb = comb[, -grep("Row.names", colnames(comb))]

for (x in 1:ncol(comb)){
  comb[,x] = as.numeric(comb[,x])
}

comb = merge(comb, pd, by.x = "row.names", by.y = "LAD4")

#build comb_rna
x = rna_files[1]
comb_rna = data.frame()
for (x in rna_files){
  cur = readRDS(x)
  names(cur) = gsub("model_", "", colnames(cur))
  cur = cur[!is.na(cur$Row.names), ]
  rownames(cur) = cur$Row.names
  comb_rna = merge(comb_rna, cur, by = "row.names", all = T)
  rownames(comb_rna) = comb_rna$Row.names
  comb_rna$Row.names = NULL
} 

comb_rna = comb_rna[, -grep("Row.names", colnames(comb_rna))]

for (x in 1:ncol(comb_rna)){
  comb_rna[,x] = as.numeric(comb_rna[,x])
}

comb_rna = merge(comb_rna, pd_rna, by.x = "row.names", by.y = "NWGC_ID")
rownames(comb_rna) = comb_rna$Row.names


dim(comb) 
dim(comb_rna) 


```

prot
```{r}
df = NULL

#fev1
mod = lm(POST_FEV1_DERV_V1 ~ protein_FEV1_post_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
null = lm(POST_FEV1_DERV_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FEV1", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#fvc
mod = lm(POST_FVC_DERV_V1 ~ protein_FVC_post_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
null = lm(POST_FVC_DERV_V1 ~  GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FVC", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#fev1/fvc
mod = lm(POST_FEV1FVC_DERV_V1 ~ protein_FEV1_FVC_post_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
null = lm(POST_FEV1FVC_DERV_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FEV1/FVC", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#GOLD
mod = lm(as.numeric(GOLD_STAGE_COPD_SEVERITY_V1) ~ protein_finalGold_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb[!is.na(comb$GOLD_STAGE_COPD_SEVERITY_V1) & comb$GOLD_STAGE_COPD_SEVERITY_V1 != "N", ])
null = lm(as.numeric(GOLD_STAGE_COPD_SEVERITY_V1) ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb[!is.na(comb$GOLD_STAGE_COPD_SEVERITY_V1) & comb$GOLD_STAGE_COPD_SEVERITY_V1 != "N", ])
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("GOLD", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#copd
comb$copd = 0
comb[!is.na(comb$GOLD_STAGE_COPD_SEVERITY_V1) & comb$GOLD_STAGE_COPD_SEVERITY_V1 != "0" & comb$GOLD_STAGE_COPD_SEVERITY_V1 != "N", ]$copd = 1

mod = glm(copd ~ protein_copd_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb, family = "binomial")
null = glm(copd ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb, family = "binomial")
terms = get_model_data(mod, type = "std")
part = NA
pred1 = predict(mod, type = "response")
pheno = comb[rownames(comb) %in% names(pred1), ]$copd
rsq.full = auc(pheno, pred1)[1]
pred2 = predict(null, type = "response")
pheno = comb[rownames(comb) %in% names(pred2), ]$copd
rsq.null = auc(pheno, pred2)[1]
df = rbind(df, c("COPD", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#POST_FEF2575_DERV
mod = lm(POST_FEF2575_DERV_V1 ~ protein_FEF2575_post_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
null = lm(POST_FEF2575_DERV_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FEF25-75", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#PEF -> POST_PEFR_DERV
mod = lm(POST_PEFR_DERV_V1 ~ protein_PEF_post_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
null = lm(POST_PEFR_DERV_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("PEF", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#6mw > SIX_MINUTE_WALK_DISTANCE_V1_A 
mod = lm(SIX_MINUTE_WALK_DISTANCE_V1_A ~ protein_distwalked_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
null = lm(SIX_MINUTE_WALK_DISTANCE_V1_A ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("6MW Distance", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#emphy -> LOG950_VIDA_V1
mod = lm(LOG950_VIDA_V1 ~ protein_pctEmph_Thirona_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
null = lm(LOG950_VIDA_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("Emphysema", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#sao2 -> SMW04A_V1
mod = lm(SMW04A_V1 ~ protein_Resting_SaO2_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
null = lm(SMW04A_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("SpO2", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 


#tlc -> BOTH_AIR_VOLUME_CM3
comb$BOTH_AIR_VOLUME_CM3 = log(comb$BOTH_AIR_VOLUME_CM3)
mod = lm(BOTH_AIR_VOLUME_CM3 ~ protein_TLC_Thirona_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
null = lm(BOTH_AIR_VOLUME_CM3 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("TLC", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 


#density. BOTH_MEDIAN_HU. 
mod = lm(BOTH_MEDIAN_HU ~ protein_lung_density_vnb_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
null = lm(BOTH_MEDIAN_HU ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("Lung Density", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#pi10.WHOLE_TREE_ALL is for all segments. 
comb$WHOLE_TREE_ALL = log(comb$WHOLE_TREE_ALL)
mod = lm(WHOLE_TREE_ALL ~ protein_Pi10_Thirona_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
null = lm(WHOLE_TREE_ALL ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("Pi10", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 


#NEGBINOM
#BODE
mod = glm.nb(BODE_INDEX_V1_A ~ protein_BODE_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
null = glm.nb(BODE_INDEX_V1_A ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("BODE", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, NA, NA)) 

#SGRQ
comb$sgrq = round(comb$SGR_TOTALSCORE_V1_A)
mod = glm.nb(sgrq ~ protein_SGRQ_scoreTotal_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
null = glm.nb(sgrq ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("SGRQ", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, NA, NA)) 

#chronbronch
mod = glm(CHRONIC_BRONCHITIS_SGR_V1 ~ protein_Chronic_Bronchitis_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb, family = "binomial")
null = glm(CHRONIC_BRONCHITIS_SGR_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb, family = "binomial")
terms = get_model_data(mod, type = "std")
part = NA
pred1 = predict(mod, type = "response")
pheno = comb[rownames(comb) %in% names(pred1), ]$CHRONIC_BRONCHITIS_SGR_V1
rsq.full = auc(pheno, pred1)[1]
pred2 = predict(null, type = "response")
pheno = comb[rownames(comb) %in% names(pred2), ]$CHRONIC_BRONCHITIS_SGR_V1
rsq.null = auc(pheno, pred2)[1]
df = rbind(df, c("Chronic Bronchitis", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 


prot_df = as.data.frame(df)
names(prot_df) = c("pheno", "pval", "beta", "ci.low", "ci.high", "rsq.partial", "rsq.null", "rsq.full")
for (x in 2:ncol(prot_df)){
  prot_df[,x] = as.numeric(prot_df[,x])
}



```


met
```{r}
df = NULL

#fev1
mod = lm(POST_FEV1_DERV_V1 ~ metabolite_FEV1_post_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
null = lm(POST_FEV1_DERV_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FEV1", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#fvc
mod = lm(POST_FVC_DERV_V1 ~ metabolite_FVC_post_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
null = lm(POST_FVC_DERV_V1 ~  GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FVC", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#fev1/fvc
mod = lm(POST_FEV1FVC_DERV_V1 ~ metabolite_FEV1_FVC_post_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
null = lm(POST_FEV1FVC_DERV_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FEV1/FVC", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#GOLD
mod = lm(as.numeric(GOLD_STAGE_COPD_SEVERITY_V1) ~ metabolite_finalGold_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb[!is.na(comb$GOLD_STAGE_COPD_SEVERITY_V1) & comb$GOLD_STAGE_COPD_SEVERITY_V1 != "N", ])
null = lm(as.numeric(GOLD_STAGE_COPD_SEVERITY_V1) ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb[!is.na(comb$GOLD_STAGE_COPD_SEVERITY_V1) & comb$GOLD_STAGE_COPD_SEVERITY_V1 != "N", ])
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("GOLD", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#copd
comb$copd = 0
comb[!is.na(comb$GOLD_STAGE_COPD_SEVERITY_V1) & comb$GOLD_STAGE_COPD_SEVERITY_V1 != "0" & comb$GOLD_STAGE_COPD_SEVERITY_V1 != "N", ]$copd = 1

mod = glm(copd ~ metabolite_copd_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb, family = "binomial")
null = glm(copd ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb, family = "binomial")
terms = get_model_data(mod, type = "std")
part = NA
pred1 = predict(mod, type = "response")
pheno = comb[rownames(comb) %in% names(pred1), ]$copd
rsq.full = auc(pheno, pred1)[1]
pred2 = predict(null, type = "response")
pheno = comb[rownames(comb) %in% names(pred2), ]$copd
rsq.null = auc(pheno, pred2)[1]
df = rbind(df, c("COPD", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#POST_FEF2575_DERV
mod = lm(POST_FEF2575_DERV_V1 ~ metabolite_FEF2575_post_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
null = lm(POST_FEF2575_DERV_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FEF25-75", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#PEF -> POST_PEFR_DERV
mod = lm(POST_PEFR_DERV_V1 ~ metabolite_PEF_post_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
null = lm(POST_PEFR_DERV_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + HT_CM_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("PEF", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#6mw SIX_MINUTE_WALK_DISTANCE_V1_A 
mod = lm(SIX_MINUTE_WALK_DISTANCE_V1_A ~ metabolite_distwalked_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
null = lm(SIX_MINUTE_WALK_DISTANCE_V1_A ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("6MW Distance", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#emphy -> LOG950_VIDA_V1
mod = lm(LOG950_VIDA_V1 ~ metabolite_pctEmph_Thirona_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
null = lm(LOG950_VIDA_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("Emphysema", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#sao2. SMW04A_V1
mod = lm(SMW04A_V1 ~ metabolite_Resting_SaO2_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
null = lm(SMW04A_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("SpO2", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 


#tlc.  BOTH_AIR_VOLUME_CM3
comb$BOTH_AIR_VOLUME_CM3 = log(comb$BOTH_AIR_VOLUME_CM3)
mod = lm(BOTH_AIR_VOLUME_CM3 ~ metabolite_TLC_Thirona_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
null = lm(BOTH_AIR_VOLUME_CM3 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("TLC", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 


#density. BOTH_MEDIAN_HU. 
mod = lm(BOTH_MEDIAN_HU ~ metabolite_lung_density_vnb_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
null = lm(BOTH_MEDIAN_HU ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("Lung Density", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#pi10.  WHOLE_TREE_ALL 
comb$WHOLE_TREE_ALL = log(comb$WHOLE_TREE_ALL)
mod = lm(WHOLE_TREE_ALL ~ metabolite_Pi10_Thirona_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
null = lm(WHOLE_TREE_ALL ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1 + MODEL, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("Pi10", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#NEGBINOM
#BODE
mod = glm.nb(BODE_INDEX_V1_A ~ metabolite_BODE_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
null = glm.nb(BODE_INDEX_V1_A ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("BODE", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, NA, NA)) 

#SGRQ
comb$sgrq = round(comb$SGR_TOTALSCORE_V1_A)
mod = glm.nb(sgrq ~ metabolite_SGRQ_scoreTotal_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
null = glm.nb(sgrq ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("SGRQ", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, NA, NA)) 

#chronbronch
mod = glm(CHRONIC_BRONCHITIS_SGR_V1 ~ metabolite_Chronic_Bronchitis_P2 + GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb, family = "binomial")
null = glm(CHRONIC_BRONCHITIS_SGR_V1 ~ GENDER + as.factor(RACE) + AGE_DERV_V1 + CURRENT_SMOKER_V1 + SMOKING_PACK_YEARS_V1, comb, family = "binomial")
terms = get_model_data(mod, type = "std")
part = NA
pred1 = predict(mod, type = "response")
pheno = comb[rownames(comb) %in% names(pred1), ]$CHRONIC_BRONCHITIS_SGR_V1
rsq.full = auc(pheno, pred1)[1]
pred2 = predict(null, type = "response")
pheno = comb[rownames(comb) %in% names(pred2), ]$CHRONIC_BRONCHITIS_SGR_V1
rsq.null = auc(pheno, pred2)[1]
df = rbind(df, c("Chronic Bronchitis", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 


met_df = as.data.frame(df)
names(met_df) = c("pheno", "pval", "beta", "ci.low", "ci.high", "rsq.partial", "rsq.null", "rsq.full")
for (x in 2:ncol(met_df)){
  met_df[,x] = as.numeric(met_df[,x])
}


```

trans
```{r}
df = NULL

#fev1
mod = lm(POST_FEV1_DERV ~ transcript_FEV1_post_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS + HT_CM, comb_rna)
null = lm(POST_FEV1_DERV ~ GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS + HT_CM, comb_rna)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FEV1", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#fvc
mod = lm(POST_FVC_DERV ~ transcript_FVC_post_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS + HT_CM, comb_rna)
null = lm(POST_FVC_DERV ~  GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS + HT_CM, comb_rna)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FVC", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#fev1/fvc
mod = lm(POST_FEV1FVC_DERV ~ transcript_FEV1_FVC_post_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS + HT_CM, comb_rna)
null = lm(POST_FEV1FVC_DERV ~ GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS + HT_CM, comb_rna)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FEV1/FVC", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#GOLD 
mod = lm(as.numeric(GOLD_STAGE_COPD_SEVERITY) ~ transcript_finalGold_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna[!is.na(comb_rna$GOLD_STAGE_COPD_SEVERITY) & comb_rna$GOLD_STAGE_COPD_SEVERITY != "N", ])
null = lm(as.numeric(GOLD_STAGE_COPD_SEVERITY) ~ GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna[!is.na(comb_rna$GOLD_STAGE_COPD_SEVERITY) & comb_rna$GOLD_STAGE_COPD_SEVERITY != "N", ])
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("GOLD", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#copd
comb_rna$copd = 0
comb_rna[!is.na(comb_rna$GOLD_STAGE_COPD_SEVERITY) & comb_rna$GOLD_STAGE_COPD_SEVERITY != "0" & comb_rna$GOLD_STAGE_COPD_SEVERITY != "N", ]$copd = 1

mod = glm(copd ~ transcript_copd_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna, family = "binomial")
null = glm(copd ~ GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna, family = "binomial")
terms = get_model_data(mod, type = "std")
part = NA
pred1 = predict(mod, type = "response")
pheno = comb_rna[rownames(comb_rna) %in% names(pred1), ]$copd
rsq.full = auc(pheno, pred1)[1]
pred2 = predict(null, type = "response")
pheno = comb_rna[rownames(comb_rna) %in% names(pred2), ]$copd
rsq.null = auc(pheno, pred2)[1]
df = rbind(df, c("COPD", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#POST_FEF2575_DERV
mod = lm(POST_FEF2575_DERV ~ transcript_FEF2575_post_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS + HT_CM, comb_rna)
null = lm(POST_FEF2575_DERV ~ GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS + HT_CM, comb_rna)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("FEF25-75", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#PEF -> POST_PEFR_DERV
mod = lm(POST_PEFR_DERV ~ transcript_PEF_post_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS + HT_CM, comb_rna)
null = lm(POST_PEFR_DERV ~ GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS + HT_CM, comb_rna)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("PEF", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 

#6mw.  SIX_MINUTE_WALK_DISTANCE_A 
mod = lm(SIX_MINUTE_WALK_DISTANCE ~ transcript_distwalked_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna)
null = lm(SIX_MINUTE_WALK_DISTANCE ~ GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("6MW Distance", summary(mod)$coefficients[2, "Pr(>|t|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 


#NEGBINOM
#BODE
mod = glm.nb(BODE_INDEX ~ transcript_BODE_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna)
null = glm.nb(BODE_INDEX ~ GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("BODE", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, NA, NA)) 

#SGRQ
comb_rna$sgrq = round(comb_rna$SGR_TOTALSCORE)
mod = glm.nb(sgrq ~ transcript_SGRQ_scoreTotal_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna)
null = glm.nb(sgrq ~ GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna)
terms = get_model_data(mod, type = "std")
part = rsq.partial(mod, null, adj = T)$partial.rsq
rsq.full = summary(mod)$adj.r.squared
rsq.null = summary(null)$adj.r.squared
df = rbind(df, c("SGRQ", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, NA, NA)) 

#chronbronch
mod = glm(CHRONIC_BRONCHITIS_SGR ~ transcript_Chronic_Bronchitis_P2 + GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna, family = "binomial")
null = glm(CHRONIC_BRONCHITIS_SGR ~ GENDER + as.factor(RACE) + AGE_DERV + CURRENT_SMOKER + SMOKING_PACK_YEARS, comb_rna, family = "binomial")
terms = get_model_data(mod, type = "std")
part = NA
pred1 = predict(mod, type = "response")
pheno = comb_rna[rownames(comb_rna) %in% names(pred1), ]$CHRONIC_BRONCHITIS_SGR
rsq.full = auc(pheno, pred1)[1]
pred2 = predict(null, type = "response")
pheno = comb_rna[rownames(comb_rna) %in% names(pred2), ]$CHRONIC_BRONCHITIS_SGR
rsq.null = auc(pheno, pred2)[1]
df = rbind(df, c("Chronic Bronchitis", summary(mod)$coefficients[2, "Pr(>|z|)"], terms$estimate[1], terms$conf.low[1], terms$conf.high[1], part, rsq.null, rsq.full)) 


trans_df = as.data.frame(df)
names(trans_df) = c("pheno", "pval", "beta", "ci.low", "ci.high", "rsq.partial", "rsq.null", "rsq.full")
for (x in 2:ncol(trans_df)){
  trans_df[,x] = as.numeric(trans_df[,x])
}

```

combine
```{r}
prot_df$type = "Protein"
met_df$type = "Metabolite"
trans_df$type = "Transcript"
df = as.data.frame(rbind(prot_df, met_df, trans_df))

df$bh = p.adjust(df$pval, method = "BH")
df$sig = F
df[df$bh <= .05, ]$sig = T

df$class = "quant"
df[df$pheno == "COPD" | df$pheno == "Chronic Bronchitis", ]$class = "binary"
df[df$pheno == "SGRQ" | df$pheno == "BODE", ]$class = "negbinom"


df$class2 = df$class
df[df$class2 == "negbinom" | df$class2 == "binary", ]$class2 = "Binomial"
df[df$class2 == "quant", ]$class2 = "Gaussian"

df$class2 = factor(df$class2, levels = c("Gaussian", "Binomial"))

write.csv(df, "C:/Users/konigsbi/Desktop/ors_topmed/March/Full_SPIROMICS_forSupp.csv")

a_two = ggplot(df, aes(beta, -log10(bh), color = type)) + 
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", linewidth = 0.4) +
  geom_point() +
  ggrepel::geom_label_repel(aes(label = pheno, fill = type), alpha = 0.7, size = 3, box.padding = 0.1, seed = 11, color = "black") +
  facet_wrap(~class2, scales = "free") +
  labs(x = "Effect Size", y = "-log10(adjusted p-value)", title = "SPIROMICS Replication", color = "") + 
  theme(strip.background = element_blank(), strip.text.x = element_text(size = 10)) +
  guides(fill = "none")
#a_two
```


multipanel rep fig
```{r eval = F}
a = ggplot(df[df$class == "quant", ], aes(beta, factor(pheno, levels = rev(levels(factor(pheno)))), alpha = sig, group = type, color = type)) + 
  geom_point(position = position_dodge(width = .5)) + 
  geom_vline(xintercept = 0, linetype = "dashed", color = "grey30") + 
  geom_errorbar(aes(xmin = ci.low, xmax = ci.high), width = 0, position = position_dodge(width = .5)) + 
  theme(plot.title = element_text(hjust = 0.5), legend.position = "none") + 
  labs(x = "\u03b2 Coefficient", y = "") + 
  scale_alpha_manual(values = c(1)) +
  guides(alpha = "none")

#OR
b = ggplot(df[df$class == "binary", ], aes(beta, factor(pheno, levels = rev(levels(factor(pheno)))), alpha = sig, group = type, color = type)) + 
  geom_point(position = position_dodge(width = .5)) + 
  geom_vline(xintercept = 1, linetype = "dashed", color = "grey30") + 
  geom_errorbar(aes(xmin = ci.low, xmax = ci.high), width = 0, position = position_dodge(width = .5)) + 
  theme(plot.title = element_text(hjust = 0.5)) + 
  labs(x = "Odds Ratio", y = "") + 
  scale_alpha_manual(values = c(1)) +
  guides(alpha = "none")

#plot rsqs
c = ggplot(df[df$class == "quant", ]) + 
  geom_bar(aes(rsq.full, pheno, fill = type, alpha = sig), stat = "identity", position = "dodge") + 
  geom_bar(aes(rsq.null, pheno), stat = "identity", position = "dodge") + 
  theme(plot.title = element_text(hjust = 0.5), legend.position = "none") + 
  labs(x = bquote("Adjusted R"^2), y = "", fill = "") +
  scale_y_discrete(limits = rev(levels(factor(df[df$class == "quant", ]$pheno)))) + 
  scale_alpha_manual(values = c(1)) +
  guides(alpha = "none") +
  xlim(0, 1)

#AUC
d = ggplot(df[df$class == "binary", ]) + 
  geom_bar(aes(rsq.full, factor(pheno, levels = rev(levels(factor(pheno)))), fill = type, alpha = sig), stat = "identity", position = "dodge") + 
  geom_bar(aes(rsq.null, pheno), stat = "identity", position = "dodge") + 
  theme(plot.title = element_text(hjust = 0.5)) + 
  labs(x = bquote("AUROC"), y = "", fill = "") +
  #scale_y_discrete(limits = rev(levels(factor(df[df$class == "quant", ]$pheno)))) + 
  scale_alpha_manual(values = c(1)) +
  guides(alpha = "none") +
  coord_cartesian(xlim = c(0.5, 1))

png("C:/Users/konigsbi/Desktop/ors_topmed/March/Figures/SPIROMICS_replication.png")
cowplot::plot_grid(a, b, c, d, labels = c("a.", "b.", "c.", "d."))
dev.off()
```

# Figure 3 

```{r}
#load MESA
load("C:/Users/konigsbi/Desktop/ors_topmed/March/Figures/mesa_rep.Rdata")

a2 = ggplot(df[df$class == "quant", ], aes(beta, factor(pheno, levels = rev(levels(factor(pheno)))), alpha = sig, group = type, color = type)) + 
  geom_point(position = position_dodge(width = .5)) + 
  geom_vline(xintercept = 0, linetype = "dashed", color = "grey30") + 
  geom_errorbar(aes(xmin = ci.low, xmax = ci.high), width = 0, position = position_dodge(width = .5)) + 
  theme(plot.title = element_text(hjust = 0.5), legend.position = "none") + 
  labs(x = "\u03b2 Coefficient", y = "") + 
  scale_alpha_manual(values = c(1)) +
  guides(alpha = "none")

#OR
b2 = ggplot(df[df$class == "binary", ], aes(beta, factor(pheno, levels = rev(levels(factor(pheno)))), alpha = sig, group = type, color = type)) + 
  geom_point(position = position_dodge(width = .5)) + 
  geom_vline(xintercept = 1, linetype = "dashed", color = "grey30") + 
  geom_errorbar(aes(xmin = ci.low, xmax = ci.high), width = 0, position = position_dodge(width = .5)) + 
  theme(plot.title = element_text(hjust = 0.5), legend.position = "none") + 
  labs(x = "Odds Ratio", y = "") + 
  scale_alpha_manual(values = c(1)) +
  guides(alpha = "none")

#plot rsqs
c2 = ggplot(df[df$class == "quant", ]) + 
  geom_bar(aes(rsq.full, pheno, fill = type, alpha = sig), stat = "identity", position = "dodge") + 
  geom_bar(aes(rsq.null, pheno), stat = "identity", position = "dodge") + 
  theme(plot.title = element_text(hjust = 0.5), legend.position = "none") + 
  labs(x = bquote("Adjusted R"^2), y = "", fill = "") +
  scale_y_discrete(limits = rev(levels(factor(df[df$class == "quant", ]$pheno)))) + 
  scale_alpha_manual(values = c(1)) +
  guides(alpha = "none") +
  xlim(0, 1)

#AUC
d2 = ggplot(df[df$class == "binary", ]) + 
  geom_bar(aes(rsq.full, pheno, fill = type, alpha = sig), stat = "identity", position = "dodge") + 
  geom_bar(aes(rsq.null, pheno), stat = "identity", position = "dodge") + 
  theme(plot.title = element_text(hjust = 0.5), legend.position = "none") + 
  labs(x = bquote("AUROC"), y = "", fill = "") +
  #scale_y_discrete(limits = rev(levels(factor(df[df$class == "quant", ]$pheno)))) + 
  scale_alpha_manual(values = c(1)) +
  guides(alpha = "none") +
  xlim(0, 1)

png("C:/Users/konigsbi/Desktop/ors_topmed/March/Figures/Figure3.png", units = "in", res = 500, height = 15, width = 12)
one = cowplot::plot_grid(a_two, labels = c("a. "))
two = cowplot::plot_grid(a2, b2, c2, d2, nrow = 1, labels = c("b. ", "c. ", "d. ", "e. "))
three = cowplot::plot_grid(a_one, labels = c("f. "))
four = cowplot::plot_grid(a, b, c, d, nrow = 1, labels = c("g. ", "h. ", "i. ", "j. "))
cowplot::plot_grid(one, two, three, four, nrow = 4, rel_heights = c(1, .8, 1, .8))
dev.off()

```

***

# Wrap-Up

```{r}
Sys.time() - start

sessionInfo()

```
